net/NetUtil.cpp

```diff
@@ -138,6 +154,88 @@ bool isLocalhost(const std::string& targetHost)
     return false;
 }
 
+void AsyncDNS::startThread()
+{
+    assert(!_thread);
+    _exit = false;
+    _thread.reset(new std::thread(&AsyncDNS::resolveDNS, this));
+}
+
+void AsyncDNS::joinThread()
+{
+    _exit = true;
+    _condition.notify_all();
+    _thread->join();
+    _thread.reset();
+}
+
+AsyncDNS::AsyncDNS()
+    : _resolver(std::make_unique<DNSResolver>())
+{
+    startThread();
+}
+
+AsyncDNS::~AsyncDNS()
+{
+    joinThread();
+}
+
+void AsyncDNS::resolveDNS()
+{
+    std::unique_lock<std::mutex> guard(_lock);
+    while (true)
+    {
+        while (_lookups.empty() && !_exit)
+            _condition.wait(guard);
+
+        if (_exit)
+            break;
+
+        Lookup lookup = _lookups.front();
+        _lookups.pop();
+        std::string hostToCheck, exception;
+
+        try
+        {
+            hostToCheck = _resolver->resolveDNS(lookup.query).name();
+        }
+        catch (const Poco::Exception& exc)
+        {
+            exception = "net::canonicalHostName(\"" + lookup.query + "\") failed: " + exc.displayText();
+        }
+
+        lookup.cb(hostToCheck, exception);
+    }
+}
+
+void AsyncDNS::addLookup(const std::string& lookup, const DNSThreadFn& cb)
+{
+    std::unique_lock<std::mutex> guard(_lock);
```

Comment:
Given the above - superficially it looks like this will still block the main thread while trying to push a lookup if the other thread is doing a slow sync. lookup (?) - which might defeat the parallelism surely ?